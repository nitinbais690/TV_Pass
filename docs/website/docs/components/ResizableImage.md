---
id: ResizableImage
title: ResizableImage
---

import useBaseUrl from '@docusaurus/useBaseUrl';

A custom Image component that wraps default `Image` component for displaying network images
and works with Firstlight's ImageResizer server by generating a deterministic image url with specific resizing params.

:::tip
Accepts standard `ImageProps` along with width and height optionally.
:::

:::note
`ResizableImage` supports computing the dimensions when the view gets layed-out (onLayout event),
this should be in layouts, while this is convenient in many layouts, we would recommend passing in
width and height props while using `ResizableImage` within a list view
since computing the width and height onLayout can be very expensive on a complicated list,
which would lead to stuttering on scroll.
:::

## Usage

```jsx
import { ResizableImage } from 'qp-discovery-ui';

<ResizableImage
    keyId={'123'}
    imageType={ImageType.Poster}
    aspectRatioKey={AspectRatio._16by9}
/>;
```

#### ResizableImage with custom image styles

```jsx {10}
import { ResizableImage } from 'qp-discovery-ui';

const style = StyleSheet.create({
    margin: 10,
    backgroundColor: 'black',
});

<ResizableImage
    keyId={'123'}
    style={style}
    imageType={ImageType.Poster}
    aspectRatioKey={AspectRatio._16by9}
/>;
```

---

## Props

-   [`keyId`](#keyId)
-   [`width`](#width)
-   [`height`](#height)
-   [`testID`](#testID)
-   [`skipResize`](#skipResize)
-   [`aspectRatioKey`](#aspectRatioKey)
-   [`imageType`](#imageType)

---

## Reference

### `keyId`

The resourceId for which to fetch favorite

|  Type  | Default |
| :----: | :-----: |
| string |    -    |

---

### `width`

The width of the image (Optional).

|  Type  | Default |
| :----: | :-----: |
| number |  none   |

---

### `height`

The height of the image (Optional).

|  Type  | Default |
| :----: | :-----: |
| number |  none   |

---

### `testID`

Unique identifier to identify the element in tests

|  Type  | Default |
| :----: | :-----: |
| string |  true   |

---

### `skipResize`

Optionally skips passing resize params.

|  Type   | Default |
| :-----: | :-----: |
| boolean |  false  |

---

### `aspectRatioKey`

The aspect ratio of the image.

|     Type      | Default |
| :-----------: | :-----: |
| `AspectRatio` |  none   |

---

### `imageType`

The type of the image.

|    Type     | Default |
| :---------: | :-----: |
| `ImageType` |  none   |

---
