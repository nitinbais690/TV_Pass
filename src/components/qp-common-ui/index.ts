export * from './src/styles/responsiveUtils';
export * from './src/styles/stylesheet';
export * from './src/styles/typography';
