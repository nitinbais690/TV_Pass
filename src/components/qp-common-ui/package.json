{
    "name": "qp-common-ui",
    "version": "0.0.1",
    "description": "A collection of Firstlight's common UI Components",
    "main": "index.ts",
    "scripts": {
        "test": "echo \"Error: no test specified\" && exit 1"
    },
    "keywords": [
        "qp-common-ui"
    ],
    "private": true,
    "peerDependencies": {
        "@typescript-eslint/eslint-plugin": "^1.9.0",
        "@typescript-eslint/parser": "^1.9.0",
        "eslint": "^5.16.0",
        "eslint-config-prettier": "^4.3.0",
        "eslint-plugin-prettier": "^3.1.0",
        "prettier": "^1.17.1",
        "react-native": ">=0.61.3",
        "react-native-device-info": ">=4.0.2-rc.1"
    },
    "devDependencies": {
        "@types/react-native": "0.57.57",
        "typescript": "3.4.5"
    }
}
