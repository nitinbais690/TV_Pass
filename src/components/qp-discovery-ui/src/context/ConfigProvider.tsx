import React, { useContext } from 'react';
import { ServiceConfig, createAPIClient } from '../api';
import { ClientContextProvider, Client } from 'react-fetching-library';
import { APIHandler } from 'core/api/api-handler';
type ConfigProviderProps = { config: Config; serviceConfig: ServiceConfig; children: React.ReactNode };

export type Config = { [key: string]: any };

const ConfigContext = React.createContext<Config | null>(null);

export const ConfigProvider = ({ config, serviceConfig, children }: ConfigProviderProps) => {
    const client: Client = createAPIClient(serviceConfig);
    APIHandler.getInstance().setClient(client); //To perform api without useQuery
    return (
        <ConfigContext.Provider value={config}>
            <ClientContextProvider client={client}>{children}</ClientContextProvider>
        </ConfigContext.Provider>
    );
};

export const useConfig = () => {
    const context = useContext(ConfigContext);
    if (context === undefined) {
        throw new Error('useConfig must be used within a ConfigProvider');
    }
    return context;
};
