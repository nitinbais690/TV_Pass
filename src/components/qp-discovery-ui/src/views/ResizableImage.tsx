import React, { useState } from 'react';

import { ImageProps, LayoutChangeEvent, ImageURISource } from 'react-native';
import FastImage from 'react-native-fast-image';
import { useConfig } from '../context/ConfigProvider';
import { AspectRatio, ImageType } from 'qp-common-ui';
import { imageResizerUri, imageResizer } from '../utils/ImageUtils';

interface ResizableImageProps extends Partial<ImageProps> {
    keyId: string;
    width?: number;
    height?: number;
    testID?: string;
    skipResize?: boolean;
    isPortrait?: boolean;
    aspectRatioKey: AspectRatio;
    imageType: ImageType;
}

/**
 * A Custom React component that wraps `Image` component for displaying network images
 * and works with Firstlight's ImageResizer server by dynamically adding required parameters.
 *
 * @param  {ResizableImageProps} props Takes standards `ImageProps` along with width and height optionally.
 *
 * @note `ResizableImage` supports computing the dimensions when the view gets layed out (onLayout event),
 * this should be in layouts, while this is convenient in many layouts, we would recommend passing in
 * width and height props while using `ResizableImage` within a list view
 * since computing the width and height onLayout can be very expensive on a complicated list,
 * which would lead to stuttering on scroll.
 */
const ResizableImage = (props: ResizableImageProps): JSX.Element | null => {
    const config = useConfig();
    const resizerEndpoint = (config && config.imageResizeURL) || undefined;
    const resizerPath = /*(config && config.imageResizerPath)*/ 'image' || undefined;
    let [uri, setUri] = useState(props.source ? (props.source as ImageURISource).uri : '');
    let onLayout = (_: LayoutChangeEvent): void => {};

    if (props.keyId.length < 1 && props.aspectRatioKey) {
        return null;
    }

    if (!uri && resizerEndpoint && !props.skipResize) {
        if (props.width) {
            uri = imageResizerUri(
                resizerEndpoint,
                resizerPath,
                props.keyId,
                props.aspectRatioKey,
                props.imageType,
                props.width,
            );
        } else {
            onLayout = (e: LayoutChangeEvent): void => {
                const l = e.nativeEvent.layout;
                setUri(
                    imageResizerUri(
                        resizerEndpoint,
                        resizerPath,
                        props.keyId,
                        props.aspectRatioKey,
                        props.imageType,
                        l.width,
                    ),
                );
            };
        }
    } else {
        if (props.width && props.height && !props.skipResize) {
            uri = imageResizer(uri, props.width);
        } else {
            onLayout = (e: LayoutChangeEvent): void => {
                const l = e.nativeEvent.layout;
                setUri(imageResizer(uri, l.width));
            };
        }
    }
    return (
        <FastImage
            testID={props.testID}
            style={props.style}
            source={{ uri: uri, cache: FastImage && FastImage.cacheControl ? FastImage.cacheControl.web : undefined }}
            onLayout={onLayout}
            resizeMode={FastImage && FastImage.resizeMode ? FastImage.resizeMode.contain : undefined}
        />
    );
};

// export default ResizableImage;
export default React.memo(ResizableImage, (prev: ResizableImageProps, next: ResizableImageProps): boolean => {
    return (
        prev.keyId === next.keyId && prev.isPortrait === next.isPortrait && prev.aspectRatioKey === next.aspectRatioKey
    );
});
