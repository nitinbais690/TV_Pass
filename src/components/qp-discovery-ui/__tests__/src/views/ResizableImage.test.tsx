import React from 'react';
import ResizableImage from '../../../src/views/ResizableImage';
import { ConfigProvider } from '../../../src/context/ConfigProvider';

import { render, toJSON, fireEvent, NativeTestEvent } from '@testing-library/react-native';
import { AspectRatio, ImageType } from 'qp-common-ui';

jest.mock('react-native-fast-image', () => {
    return require('react-native').Image;
});

describe('ResizableImage', () => {
    const imageType = ImageType.Poster;
    const aspectRatioKey = AspectRatio._1by1;
    it('renders correctly', () => {
        const { getByTestId } = render(
            <ResizableImage testID={'image'} keyId="01234" imageType={imageType} aspectRatioKey={aspectRatioKey} />,
        );
        expect(getByTestId('image')).not.toBeNull();
    });

    it('renders nothing when an invalid uri is passed', () => {
        const { queryByTestId } = render(
            <ResizableImage testID={'image'} keyId={''} imageType={imageType} aspectRatioKey={aspectRatioKey} />,
        );
        expect(queryByTestId('image')).toBeNull();
    });

    it('renders correctly when no dimensions are passed in', () => {
        const config = {
            imageResizeURL: 'https://image-resizer-service.apisbox.devops.quickplay.com',
            // imageResizerPath: 'image',
        };

        const { getByTestId } = render(
            <ConfigProvider config={config} serviceConfig={{}}>
                <ResizableImage testID={'image'} keyId="012345" imageType={imageType} aspectRatioKey={aspectRatioKey} />
            </ConfigProvider>,
        );

        const image = getByTestId('image');
        fireEvent(
            image,
            new NativeTestEvent('layout', {
                nativeEvent: {
                    layout: {
                        width: 200,
                        height: 300,
                    },
                },
            }),
        );

        expect(image.props.source.uri).toBe(
            'https://image-resizer-service.apisbox.devops.quickplay.com/image/012345/0-1x1.jpg?width=400',
        );
    });

    it('avoid re-renders correctly', () => {
        const jsx = (
            <ResizableImage testID={'image'} keyId="01234" imageType={imageType} aspectRatioKey={aspectRatioKey} />
        );
        const { getByTestId, rerender } = render(jsx);
        expect(getByTestId('image')).not.toBeNull();
        rerender(jsx);
        expect(getByTestId('image')).not.toBeNull();
    });

    it('avoid renders overridden style correctly', () => {
        const style = { backgroundColor: 'red' };
        const { container } = render(
            <ResizableImage
                testID={'image'}
                keyId="01234"
                imageType={imageType}
                aspectRatioKey={aspectRatioKey}
                style={style}
            />,
        );
        expect(toJSON(container)).toMatchSnapshot();
    });

    it('renders correctly when dimensions are passed in', () => {
        const { container } = render(
            <ResizableImage
                testID={'image'}
                keyId="sdsd"
                width={300}
                height={300}
                imageType={imageType}
                aspectRatioKey={aspectRatioKey}
            />,
        );
        expect(toJSON(container)).toMatchSnapshot();
    });
});
