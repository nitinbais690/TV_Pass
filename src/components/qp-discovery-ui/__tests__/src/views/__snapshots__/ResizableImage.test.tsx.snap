// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResizableImage avoid renders overridden style correctly 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Image
    source={
      Object {
        "cache": undefined,
        "uri": "",
      }
    }
    style={
      Object {
        "backgroundColor": "red",
      }
    }
    testID="image"
  />
</View>
`;

exports[`ResizableImage renders correctly when dimensions are passed in 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Image
    source={
      Object {
        "cache": undefined,
        "uri": "&width=600&height=600&scalingMode=aspectFit",
      }
    }
    testID="image"
  />
</View>
`;
